shared/scripts/virtio_console_guest.py

```diff
@@ -1425,6 +1425,7 @@ def daemonize(self):
         so = open(self.stdout, 'w')
         se = open(self.stderr, 'w')
 
+        time.sleep(0.1)
```

Comment:
the parent process is exited in L1417, so how do we observe this conflict ?

Reply:
@balamuruhans Sorry to confuse you, it forks twice, so the parent here I mean is actually the parent of its parent, which `return False` in L1404 and continue in run() function.
So I have got this in my debugging:
```
2018-08-13 23:32:47: Return false for parent, children pid: 1670
2018-08-13 23:32:47: Exit parent, children pid: 1672
2018-08-13 23:32:47: open files success, pid is: 1672
2018-08-13 23:32:47: dup stdin success, returned fd2 is: None, pid: 1672
2018-08-13 23:32:47: pid: 1669, file: /tmp/guest_daemon_pi
2018-08-13 23:32:47: pid: 1669, file: /tmp/guest_daemon_po
2018-08-13 23:32:47: pid: 1669, file: /tmp/guest_daemon_pe
2018-08-13 23:32:47: pid: 1672, file: /tmp/guest_daemon_pi
2018-08-13 23:32:47: Traceback (most recent call last):
2018-08-13 23:32:47:   File "/tmp/virtio_console_guest.py", line 1591, in <module>
2018-08-13 23:32:47:     main()
2018-08-13 23:32:47:   File "/tmp/virtio_console_guest.py", line 1539, in main
2018-08-13 23:32:47:     pid = list(pid)[0][0]
2018-08-13 23:32:47: IndexError: list index out of range
```

And after sleep(0.1)

```
2018-08-14 02:13:56: Return false for parent, children pid: 1628
2018-08-14 02:13:56: main, pid is: 1627
2018-08-14 02:13:56: Exit parent, children pid: 1630
2018-08-14 02:13:56: open files success, pid is: 1630
2018-08-14 02:13:56: children 1630 sleep for a while
2018-08-14 02:13:56: pid: 1627, file: /tmp/guest_daemon_pi, current pid: 1627
2018-08-14 02:13:56: pid: 1627, file: /tmp/guest_daemon_po, current pid: 1627
2018-08-14 02:13:56: pid: 1627, file: /tmp/guest_daemon_pe, current pid: 1627
2018-08-14 02:13:56: pid: 1630, file: /tmp/guest_daemon_pi, current pid: 1627
2018-08-14 02:13:56: pid: 1630, file: /tmp/guest_daemon_po, current pid: 1627
2018-08-14 02:13:56: pid: 1630, file: /tmp/guest_daemon_pe, current pid: 1627
2018-08-14 02:13:56: PASS: main Start, pid is: 1627
2018-08-14 02:13:57: dup stdin success, returned fd2 is: None, pid: 1630
2018-08-14 02:13:57: dup stdout success, returned fd2 is: None, pid: 1630
2018-08-14 02:13:57: dup stderr success, returned fd2 is: None, pid: 1630
```

Reply:
@vivianQizhu Thanks for the nice explanation, 

can we do something using flag like this, It will make the fix more robust. Please let me know your thoughts

```patch
diff --git a/shared/scripts/virtio_console_guest.py b/shared/scripts/virtio_console_guest.py
index b14d4a9c..c53ecf40 100755
--- a/shared/scripts/virtio_console_guest.py
+++ b/shared/scripts/virtio_console_guest.py
@@ -1363,12 +1363,17 @@ class Daemon:
         self.stdout = stdout
         self.stderr = stderr
 
+        # Flag to check is_file_open() completes
+        self.check_is_file_open = False
+
     @staticmethod
-    def is_file_open(path):
+    def is_file_open(path, check=False):
         """
         Determine process which open file.
 
         :param path: Path to file.
+        :param check: if True toggles flag check_is_file_open to True
+
         :return: [[pid,mode], ... ].
         """
         opens = []
@@ -1393,6 +1398,10 @@ class Daemon:
                 if e.errno == 2:
                     continue
                 raise
+
+        if check and not self.check_is_file_open:
+            self.check_is_file_open = True
+
         return opens
 
     def daemonize(self):
@@ -1425,17 +1434,23 @@ class Daemon:
         so = open(self.stdout, 'w')
         se = open(self.stderr, 'w')
 
-        os.dup2(si.fileno(), sys.stdin.fileno())
-        os.dup2(so.fileno(), sys.stdout.fileno())
-        os.dup2(se.fileno(), sys.stderr.fileno())
+        def _check_is_file_open():
+            return self.check_is_file_open
+
+        # wait to avoid race with grand father process by closing
+        # the files
+        if utils_misc.wait_for(_check_is_file_open, 60):
+            os.dup2(si.fileno(), sys.stdin.fileno())
+            os.dup2(so.fileno(), sys.stdout.fileno())
+            os.dup2(se.fileno(), sys.stderr.fileno())
 
-        si.close()
-        so.close()
-        se.close()
+            si.close()
+            so.close()
+            se.close()
 
-        sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 0)
-        sys.stderr = os.fdopen(sys.stderr.fileno(), 'w', 0)
-        return True
+            sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 0)
+            sys.stderr = os.fdopen(sys.stderr.fileno(), 'w', 0)
+            return True
 
     def start(self):
         """
@@ -1515,7 +1530,7 @@ def main():
     s_stderr = sys.stderr.fileno()
 
     pid = filter(lambda x: x[0] != str(os.getpid()),
-                 daemon.is_file_open(stdout))
+                 daemon.is_file_open(stdout, check=True))
     pid = list(pid)[0][0]
 
     print("PASS: Start")

```

Reply:
@balamuruhans Thanks for your advise, it is true that to simply sleep a while is not robust enough, but unfortunately using flag like this can not work as expected. 
Firstly is_file_open() is a `staticmethod` so no `self` object.
I tried to remove the `@staticmethod` and give it `self`, but it still not work, since the is_file_open() is executed in another process(grandfather), so the flag `self.check_is_file_open` will never being set to True(in the child process). 
And also the wait time before dup2 can not be that long, even I set it to `3`, I will hit another conflict:
```
2018-08-22 11:15:29: PASS: Start
2018-08-22 11:15:32: file open check time out, dup2 not executed
2018-08-22 11:15:32: self.check_is_file_open: False
2018-08-22 11:15:32: PASS: Daemon start.
2018-08-22 11:15:34: Traceback (most recent call last):
2018-08-22 11:15:34:   File "/tmp/virtio_console_guest.py", line 1609, in <module>
2018-08-22 11:15:34:     main()
2018-08-22 11:15:34:   File "/tmp/virtio_console_guest.py", line 1567, in main
2018-08-22 11:15:34:     os.write(d_stdin, os.read(s_stdin, 1))
2018-08-22 11:15:34: OSError: [Errno 32] Broken pipe
```
So it can not be too early and also can not be too late.
I will say that original design of this script is not robust itself.
I guess to make it strong we have to add many signals(like L1511) to ensure every steps in different process been executed in right order. But I have no confidence to schedule it very well...it is really hard to avoid dead lock if use signals between parent and children process.
So I use this sleep, it might not be a really fix, but a workaround here.

Reply:
It look the second patch fixed the issue, here is my test result:
```
# python3 virtio_console_guest.py 
Traceback (most recent call last):
  File "virtio_console_guest.py", line 1571, in <module>
    main()
  File "virtio_console_guest.py", line 1519, in main
    pid = list(pid)[0][0]
IndexError: list index out of range
```
After applied the following patch:
```patch
- sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 0)
- sys.stderr = os.fdopen(sys.stderr.fileno(), 'w', 0)
+ sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 1)
+ sys.stderr = os.fdopen(sys.stderr.fileno(), 'w', 1)
```
```
# python3 vim virtio_console_guest.py 
PASS: Start
PASS: Daemon start.
```

Would you please help double confirm this? thanks.

Reply:
@vivianQizhu, Thank you for experimenting. sorry, it was my mistake this flag logic would work on multi threaded case as it shares the same process address space. Here it will not work, let's try @luckyh's observation.

Reply:
@luckyh Thanks for help, you are right, I am removing the first commit.